### PR TITLE
Implement User, Post, Comment, and Like Models with Associated Methods

### PR DESCRIPTION
**Pull request by gaks1**
# Description:

In this pull request, we have implemented models to match the ERD diagram and established the appropriate associations. We have created **`User`**, **`Post`**, **`Comment`**, and **`Like`** models, along with their associations and custom methods to meet the project requirements.

The main updates in this PR include:

1. **Models and Associations:** We have created models for **`User`**, **`Post`**, **`Comment`**, and **`Like`**. The **`author_id`** in the **`posts`** table has been set as the foreign key for the **`users`** table. Corresponding associations have been set in each model.
2. **Data Insertion using Models:** We have tested data insertion using Rails console. Test data includes the creation of at least one user, four posts written by one of the users, and six comments on one of the posts.
3. **Custom Methods in Models:** Several custom methods have been added to the models:
    - **`User`** model includes a method **`recent_posts`** that returns the 3 most recent posts for a given user.
    - **`Post`** model includes methods **`post_counter`** and **`recent_comments`** which update the posts counter for a user and return the 5 most recent comments for a given post respectively.
    - **`Comment`** model includes a method **`comments_counter`** which updates the comments counter for a post.
    - **`Like`** model includes a method **`likes_counter`** which updates the likes counter for a post.

All the custom methods have been tested in Rails console to ensure they are working as expected.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id', counter_cache: :post_counter
+  has_many :comments
+  has_many :likes
+
+  def recent_comments
+    comments.order(created_at: :desc).limit(5)
+  end
+end
```
Comment (Whoistolu) [inline app/models/post.rb]: - [ ] [REQUIRED] Great job so far @gaks1 🥇, however, you seem to have skipped a part of the project requirement that states that the Post model should have a method that updates the posts counter for a user.
### Hint:
```ruby
def update_post_counter
  code here ...
end
```

- [ ] [OPTIONAL] I would like to also suggest that you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private.

```ruby
after_save :update_post_counter

  private

  def update_post_counter
   Code here ...
  end
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post, counter_cache: :comments_counter
+end
```
Comment (Whoistolu) [inline app/models/comment.rb]: - [ ] [REQUIRED] Also, take into consideration that a part of the project requirements states that the Comment model should have a method that updates the comments counter for a post.

### Hint:

```ruby
def update_comments_counter
  Code here ...
end
```

- [ ] [OPTIONAL] As I have recommended above, I would like to again suggest that you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comments_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private.

```ruby
after_save :update_comments_counter

  private

  def update_comments_counter
   Code here ...
  end
```

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post, counter_cache: :likes_counter
+end
```
Comment (Whoistolu) [inline app/models/like.rb]: - Let's not forget that the requirement stated that the Like model should have a method that updates the likes counter for a post.

### Hint:

```ruby
def update_likes_counter
  Code here ...
end
```

- [ ] [OPTIONAL] As I have recommended above, I would like to again suggest that you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_likes_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private.

```ruby
after_save :update_likes_counter

  private

  def update_likes_counter
   Code here ...
  end
```

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi @gaks1 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ]  _Check the inline comments under the review for sections to improve._

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Felix45 — APPROVED** [review]
# Approved :green_circle: 
Hi @gaks1 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights :heavy_check_mark: 
- [x] Rails active callbacks are used :white_check_mark: 
- [x] Model associations setup correctly :white_check_mark: 
- [x] Implemented all required changes from the previous reviewer :white_check_mark:   

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**